public/app/features/dashboard/state/PanelQueryRunner.ts

```diff
@@ -0,0 +1,184 @@
+import { getDatasourceSrv } from 'app/features/plugins/datasource_srv';
+import { Subject, Unsubscribable, PartialObserver } from 'rxjs';
+import {
+  guessFieldTypes,
+  toSeriesData,
+  PanelData,
+  LoadingState,
+  DataQuery,
+  TimeRange,
+  ScopedVars,
+  DataRequestInfo,
+  SeriesData,
+  LegacyResponseData,
+  DataQueryError,
+  toLegacyResponseData,
+  isSeriesData,
+  DataSourceApi,
+} from '@grafana/ui';
+
+import kbn from 'app/core/utils/kbn';
+
+/**
+ * Query response may come as events. any null|missing value just means
+ * that is unknown info in the request
+ */
+export interface PanelDataEvent extends Partial<PanelData> {
+  eventId: number;
+}
+
+export interface QueryRunnerOptions {
+  ds?: DataSourceApi; // if they already have the datasource, don't look it up
+  datasource: string | null;
+  queries: DataQuery[];
+  panelId: number;
+  dashboardId?: number;
+  timezone?: string;
+  timeRange?: TimeRange;
+  widthPixels: number;
+  minInterval?: string;
+  maxDataPoints?: number;
+  scopedVars?: ScopedVars;
+  cacheTimeout?: string;
+}
+
+export class PanelQueryRunner {
+  private counter = 0;
+  private subject?: Subject<PanelDataEvent>;
+  private includeLegacyFormats = false;
+
+  subscribe(observer: PartialObserver<PanelDataEvent>, includeLegacyFormats = false): Unsubscribable {
+    if (!this.subject) {
+      this.subject = new Subject();
+    }
+    if (includeLegacyFormats) {
+      this.includeLegacyFormats = true;
+    }
+    return this.subject.subscribe(observer);
+  }
+
+  async run(options: QueryRunnerOptions) {
+    if (!this.subject || !this.subject.observers.length) {
+      return; // Don't run if nobody cares!
+    }
+
+    const {
+      queries,
+      timezone,
+      datasource,
+      panelId,
+      dashboardId,
+      timeRange,
+      cacheTimeout,
+      widthPixels,
+      maxDataPoints,
+      scopedVars,
+    } = options;
+
+    const request: DataRequestInfo = {
+      timezone,
+      panelId,
+      dashboardId,
+      range: timeRange,
+      rangeRaw: timeRange.raw,
+      interval: '',
+      intervalMs: 0,
+      targets: queries,
```

Comment:
Should we deep clone queries?  This introduces a weird backdoor when panels could edit the `targets` and actually change the query.

What is grafana's general immutable strategy?  Tell people 'Don't do it'?  (i'm fine with that, but just want to be clear)

Reply:
I guess we could wrap it in an immutable proxy but I hate proxies (when debugging), and cloning feels inefficient.  But actually, cloning might be best because some data sources clones the queries already (to avoid modifying source query) and we have had a number of bugs where the data source accidentally modified the persisted query. 

So doing that cloning once here (shared clone between panels & data source) could be a good idea